File: Year 1/Semester 2/Java/Lessons/overloading_and_obj_compr/usecase_for_overloading.java
package overloading_and_obj_compr;

public class usecase_for_overloading {
    public static void main(String[] args) {
        
        Student joe = new Student("Joe");
        Student mike = new Student("Joe"); // Pay attention here
        Student dylan = new Student("Dylan");

        // System.out.println(joe == mike); false
        System.out.println(joe == mike); // This will print: false

        // The RAM sees this as correct, but we don't. We will adjust the code to work on this
        System.out.println(joe.equals(mike)); // This will print: true, because equals() is correctly overridden

        // CompareTo method
        System.out.println(joe.compareTo(mike)); // This will show 0, as both have the same name

        System.out.println(joe.compareTo(dylan)); // This will show an integer that represents the difference in lexicographical order

        // Boolean comparison
        System.out.println(joe.compareTo(mike) > 0); // This will print: false, because both names are equal
    }
}

File: Year 1/Semester 1/C/Lessons/overloading_and_obj_compr/usecase_for_overloading.c
#include <stdio.h>
#include <string.h>

// Define the Student struct
typedef struct {
    char name[50];
} Student;

// Constructor function for Student
Student createStudent(const char *name) {
    Student s;
    strcpy(s.name, name);
    return s;
}

// Equality check function (equivalent to `equals` method in Java)
int isEqual(Student s1, Student s2) {
    return strcmp(s1.name, s2.name) == 0;
}

// Comparison function (equivalent to `compareTo` method in Java)
int compareTo(Student s1, Student s2) {
    return strcmp(s1.name, s2.name);
}

int main() {
    // Creating three Student objects
    Student joe = createStudent("Joe");
    Student mike = createStudent("Joe");  // Note: Same name as joe
    Student dylan = createStudent("Dylan");

    // Pointer comparison (equivalent to '==' in Java)
    printf("joe == mike: %d\n", &joe == &mike);  // This will print: 0 (false)

    // Equality check using isEqual (equivalent to 'equals' method in Java)
    printf("joe.equals(mike): %d\n", isEqual(joe, mike));  // This will print: 1 (true)

    // Compare using compareTo (equivalent to 'compareTo' method in Java)
    printf("joe.compareTo(mike): %d\n", compareTo(joe, mike));  // This will print: 0 (names are equal)

    printf("joe.compareTo(dylan): %d\n", compareTo(joe, dylan));  // Prints comparison result based on lexicographical order

    // Boolean comparison using compareTo
    printf("joe.compareTo(mike) > 0: %d\n", compareTo(joe, mike) > 0);  // This will print: 0 (false, since the names are equal)

    return 0;
}


